_posts/2018-04-11-sign-to-contract.md

```diff
@@ -0,0 +1,400 @@
+(don't get scared, we provide a quick recap of the results at the end of the section).
+
+Given an elliptic curve with generator `G` and a second-preimage resistant hash function `h`.
+The map `m,P -> h(P||m)G + P`, where `||` stands for concatenation, `m` is the message to commit and `P` is an elliptic curve point, is a valid commitment operation.
+The output of the map is `C(m,P)`, which is an elliptic curve point.
+
+`m` is the so-called contract and `P` is a public key used for a certain (and possibly independent) purpose.
+The public key `C(m,P)` is a commitment to `m` and `P` 
+in the sense that if the input is changed, so `(m,P)!=(m',P')`, 
+then the output changes, `C(m,P)!=C(m',P')`.
+Thus we can use `C` as a commitment operation (like a hash function) inside a timestamp proof.
+
+Now, to curb the abstraction, choose `secp256k1` as elliptic curve and `sha256` as hash function. 
+The motivation is that we want to use Bitcoin as a notary, 
+and Bitcoin itself grounds its working on the integrity of these two ingredients.
+We won't use directly the corresponding `C`, but instead a slightly modified version: `OpSecp256k1Commitment`.
+It takes one input, `P||m`, and outputs the x-coord of the elliptic curve point `C(m,P)`.
+
+**Recap:** 
+pick a message to commit (also called contract) `m` and a public key `P` (on the curve `secp256k1`), concatenate them (`P||m`), then compute:
+```
+C(m,P) = SHA256(P||m)G + P 
+OpSecp256k1Commitment(P||m) = C(m,P).x
+```
+The value obtained is an elliptic curve point that serves two (independent) purposes: 
+is a public key __and__ is a commitment to `(m,P)` 
+(or only to `m`, if `P` is given before).
+
+To timestamp we need to write the x-coord of the elliptic curve point `C(m,P)` on the blockchain, 
+there are two places where such points can be found, leading to two techniques:
+- Address/Public key: *pay-to-contract*
+- Signature: *sign-to-contract*
+
+## *pay-to-contract*
+Alice has to receive some bitcoins from Bob.
+Alice has public key `A` but she also wants to timestamp a document with hash value `d`.
+Thus she tells Bob to send his bitcoins to `Q = A + h(A||d)G`.
+Bob, who eventually is not aware that he is timestamping Alice's document, 
+broadcasts the transaction, 
+which after a while becomes part of the blockchain forever.
+Alice composes her OpenTimestamps proof for her document, by finding `Q.x` on the chain and arranging the proof as follows: 
+
+```
+File sha256: d
+Timestamp:
+prepend A == d||A
+secp256k1commitment == Q.x
+prepend Q.y == Q  # y-coord is 02 or 03
+prepend TXp == TXp||Q
+append TXa == TXp||Q||TXa
+# transaction id ...
+sha256
+...
+sha256
+verify BitcoinBlockHeaderAttestation(...)
+# Bitcoin merkle root ...
+```
+So Alice has created an OTS timestamp for her document without spending a single satoshi.
+
+But, wait, has Alice received the payment from Bob?
+
+She has, but the question is delicate.
```

Comment:
I think a better way to explain the above is to just say that Alice has created a new pubkey, but it's derived based on information that she hasn't necessarily backed up already; there's no question that she's been paid.